### PR TITLE
Add implementation of Functional Neural Process

### PR DESCRIPTION
**Pull request by dereklhansen**
This adds an implementation of the Functional Neural Process (FNP) to the repo. 

In addition to the implementation of FNP, this PR currently adds two working examples of the FNP on synthetic data (located in 'case_studies/fnp':
- A one-dimensional example which illustrates different variants of the FNP
- An example on generated star-like objects which exhibit a counter-clockwise rotation

### CODE REVIEW

**Comment by dereklhansen** [comment]
This is for when the code is a bit more refactored. Torch has a built-in Multihead Attention module that might be better than the one we have copied in the code: https://pytorch.org/docs/stable/generated/torch.nn.MultiheadAttention.html
- [ ] Replace multi-head
- [ ] Replace Normal with torch.nn.Normal

**Comment by dereklhansen** [comment]
Before merging this, I decided it makes sense to delete the rotating star example. It adds a lot of lines of code to the test and doesn't actually increase coverage. Code related to convolutions can be added back as needed when we start implementing in the wake-phase.

**Comment by dereklhansen** [comment]
Also, after we get the wake phase implemented, we can delete the one-dimensional tests.

**Comment by dereklhansen** [comment]
- [ ] Remove the sklearn dependency since it's only used for the StandardScaler() in the test code

File: bliss/models/fnp.py
Comment (jeff-regier) [inline bliss/models/fnp.py:396]: Hi Derek, Would you add some brief docstrings to these classes, and re-run the makefile in the docs directory?

**Review by jeff-regier — APPROVED** [review]
Nice. This is very readable code. Good to wait for reviews from @ismael-mendoza and @Runjing-Liu120 before merging since this is such a major PR (although much more manageable now that I see how it is organized)

File: bliss/models/fnp.py
Comment (dereklhansen) [reply bliss/models/fnp.py:396]: I added the docstrings, but I don't know what I am supposed to do to regenerate the documentation

Comment (jeff-regier) [reply bliss/models/fnp.py:396]: I think it's @zzhaozhe-profolio said to type `make github` from within the `docs/docstr` directory. It's also fine to leave it for another time

Comment (dereklhansen) [reply bliss/models/fnp.py:396]: I see, I figured it out after poking around a bit. I also had to update one of the config files to have it scan the fnp.py file. 

I made the changes in the last commit on this branch, and added a small README.md file to the docs to show what to do in the future.

**Comment by jeff-regier** [comment]
Looks good to me. @ismael-mendoza @Runjing-Liu120 if you have feedback please try to weigh in sometime today

File: bliss/models/fnp.py
Comment (ismael-mendoza) [inline bliss/models/fnp.py:1]: I think you can replace all `math` functions with `numpy` functions

File: bliss/models/fnp.py
Comment (ismael-mendoza) [inline bliss/models/fnp.py:93]: We might need to change this later to work with pytorch lightning (which has separate training, validation, and testing steps). For now I think you can get rid of this additional flag using the `.training` attribute which should be a part of `nn.Module` by default.

**Review by Runjing-Liu120 — APPROVED** [review]

**Comment by Runjing-Liu120** [comment]
Looks good to me, too. The hard part will be figuring out how to add this to our wake.py / decoder.py files -- we'll have to think carefully about that. But this is great so far!

File: bliss/models/fnp.py
Comment (ismael-mendoza) [inline bliss/models/fnp.py:215]: maybe better to define this as a separate function? Not sure it has to be a muti-line lambda(hard to parse)

**Comment by ismael-mendoza** [comment]
Hi @jeff-regier and @dereklhansen , I'll be done with my review shortly

**Comment by dereklhansen** [comment]
There seems to be a problem with the CI environment that is causing the latest tests to fail. The latest commit I pushed shouldn't have caused this error - it just fixes a small error.

File: bliss/models/fnp.py
Comment (ismael-mendoza) [inline bliss/models/fnp.py:396]: Can you use [this](https://pytorch.org/docs/stable/generated/torch.nn.MultiheadAttention.html) instead of implementing from scratch? Or are you doing something sufficiently different?

File: bliss/models/fnp.py
Comment (ismael-mendoza) [inline bliss/models/fnp.py:270]: The issue you are pointing too seems to have been merged and resolved. Do you still need this separate function?

File: tests/test_fnp.py
Comment (ismael-mendoza) [inline tests/test_fnp.py:119]: extraneous comment

File: tests/test_fnp.py
Comment (ismael-mendoza) [inline tests/test_fnp.py:88]: extraneous comment

**Review by ismael-mendoza — APPROVED** [review]
Looks really good @dereklhansen, thanks for adding this code. I just had a couple of suggestions/questions.

**Comment by dereklhansen** [comment]
I added a fix for the poetry CI issues in #210

File: bliss/models/fnp.py
Comment (dereklhansen) [reply bliss/models/fnp.py]: I'm not really sure why the FNP authors linked to that issue, so I removed that comment and replaced log with log1p where appropriate.

I don't think there is a built-in to torch that does this function

File: bliss/models/fnp.py
Comment (dereklhansen) [reply bliss/models/fnp.py:388]: I moved this code over from the set_transformer repo mostly intact; maybe the MultiheadAttention got added to torch later. 

I would like to use the built-in torch module eventually, but I need to study it a bit more to make sure it is a drop-in replacement. I can open a separate issue after merging

Comment (ismael-mendoza) [reply bliss/models/fnp.py:388]: Sounds good, just wanted you to be aware this existed. My sense is that it's pretty new but not completely sure. Thanks!

**Pull request merged**
